modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/ImagePins.js

```diff
@@ -100,48 +136,231 @@ const ImagePins = ({
+			.on('drag', dragged)
+			.on('end', dragended);
+
+		const addPin = () => {
+			setCpins(
+				cPins.concat({
+					cx: 50,
+					cy: 50,
+					draggable: true,
+					fill: '#' + addNewPinState.fill,
+					id: cPins.length,
+					label: 'new' + cPins.length,
+					linked_to_sku: 'sku',
+					quantity: 0,
+					r: addNewPinState.radius,
+					sku: addNewPinState.sku,
+				})
+			);
+		};
+
+		const removePin = (id) => {
+			const currentPins = cPins.filter((element) => element.id !== id);
+
+			const newState = currentPins.map((pin, i) => {
+				return {
+					cx: pin.cx,
+					cy: pin.cy,
+					draggable: pin.draggable,
+					fill: pin.fill,
+					id: i,
+					label: pin.label,
+					linked_to_sku: pin.linked_to_sku,
+					quantity: pin.quantity,
+					r: pin.r,
+					sku: pin.sku,
+				};
+			});
+
+			setCpins(newState);
 		};
+
+		if (removePinHandler.handler) {
+			removePin(removePinHandler.pin);
+			setRemovePinHandler({
+				handler: false,
+				pin: null,
+			});
+		}
+		if (addPinHandler) {
+			setAddPinHandler(false);
+			addPin();
+		}
+
+		if (!removePinHandler.handler && !addPinHandler) {
+			try {
+				containerRef.current.selectAll('.circle_pin').remove();
+			}
+			catch (error) {
+				return;
+			}
+
+			const cont = containerRef.current
+				.selectAll('g')
+				.data(cPins)
+				.enter()
+				.append('g')
+				.attr(
+					'transform',
+					(attr) => 'translate(' + attr.cx + ',' + attr.cy + ')'
+				)
+				.attr('cx', (attr) => attr.cx)
+				.attr('cy', (attr) => attr.cy)
+				.attr('id', (attr) => attr.id)
+				.attr('label', (attr) => attr.label)
+				.attr('fill', (attr) => attr.fill)
+				.attr('linked_to_sku', (attr) => attr.linked_to_sku)
+				.attr('quantity', (attr) => attr.quantity)
+				.attr('r', (attr) => attr.r)
+				.attr('sku', (attr) => attr.sku)
+				.attr('id', (attr) => attr.id)
+				.attr('class', 'circle_pin')
+				.attr('draggable', (attr) => (attr.draggable ? true : false))
+				.call(dragHandler);
+
+			cont.append('circle')
+				.attr('r', (attr) => attr.r)
+				.attr('fill', () => '#ffffff')
+				.attr('r', (attr) => attr.r)
+				.attr('stroke', (attr) => attr.fill)
+				.attr('stroke-width', 2);
+
+			cont.append('text')
+				.text((attr) => attr.label)
+				.attr('font-size', (attr) => attr.r)
+				.attr('text-anchor', 'middle')
+				.attr('fill', '#000000')
+				.attr('alignment-baseline', 'central');
+		}
+
+		select('#newPin').on('click', handleAddPin);
+		// eslint-disable-next-line react-hooks/exhaustive-deps
 	}, [
+		addPinHandler,
+		removePinHandler,
+		changedScale,
+		cPins,
+		execZoomIn,
 		resetZoom,
 		selectedOption,
 		setResetZoom,
-		changedScale,
 		zoomOutHandler,
 		zoomInHandler,
 		enablePanZoom,
-		execZoomIn,
 		imageSettings,
 		navigationController,
 		setChangedScale,
 		setSelectedOption,
 		setZoomInHandler,
 		setZoomOutHandler,
+		handleAddPin,
+		setShowTooltip,
+		setCpins,
+		addNewPinState,
+		setRemovePinHandler,
+		setAddPinHandler,
```

Comment:
This `useLayoutEffect` is massive... Is there any way we can break it up into smaller and more manageable chunks? It's very hard to track the flow of logic and is going to be a nightmare to maintain with this set of dependencies. I worry that anyone who opens this file in the future will have a massive headache trying to fix or add anything.

Reply:
there is no way to break it because d3 works in a different way then react. so we must re-render every time diagram to make d3 do the magic

Reply:
@garofalog you can still use a single effect, but can you break it into smaller and more understandable chunks and move them outside of the effect? Or even move some of these functions outside the scope of your component. For example, [take these functions](https://github.com/liferay-frontend/liferay-portal/pull/1207/files/18e1a2602de3c424face6eb1cf3fb6b58a214431#diff-1ef8447922c18e700c4f6b6436495a12a725ff9bc175cea6d2fd94d624305d47R153-R162), they don't require anything within the scope of your component. You can move these outside of your effect so that they are not re-declaring every time and it also makes it easier to read the effect.